### ISSUE TITLE
Artifactory xray_index set to false

### ISSUE THREAD

**Issue by morgenthalep** [jfrog/terraform-provider-artifactory#557]
**Describe the bug**
We are deploying the Terraform provider to configure our remote docker repositories, but after the first apply, `xray_index` is set to `false` on every repository even though it's set to `true` in our `repos.tf`.

**Requirements for and issue**
- [x] A description of the bug
- [ ] A fully functioning terraform snippet that can be copy&pasted (no outside files or ENV vars unless that's part of the issue). **If this is not supplied, this issue will likely be closed without any effort expended.**
- [x] Your version of artifactory (you can `curl` it at `$host/artifactory/api/system/version`
  7.38.10
- [x] Your version of terraform
  1.2.8
- [x] Your version of terraform provider
  6.15.0

```tf
terraform {
  required_providers {
    artifactory = {
      source  = "registry.terraform.io/jfrog/artifactory"
      version = "6.15.0"
    }
  }
}

# Configure the Artifactory provider
provider "artifactory" {
  url          = "https://${var.artifactory_domain}/artifactory"
  access_token = "<ADMIN ACCESS TOKEN>"
}

resource "artifactory_remote_docker_repository" "remote_docker_hub" {
  key                            = "docker-hub-remote"
  url                            = "https://registry-1.docker.io/"
  notes                          = "test"
  username                       = "<USERNAME>"
  password                       = "<PASSWORD>"
  includes_pattern               = "lots of images"
  retrieval_cache_period_seconds = 43200
  missed_cache_period_seconds    = 7200
  property_sets = [
    "artifactory"
  ]
  proxy      = "<PROXY DOMAIN>"
  xray_index = true
}
```

**Expected behavior**
After applying changes, `xray_index` should be enabled, and no diff should show up where `xray_index` is re-enabled.

**Additional context**
The repositories already existed before this point, and were manually configured.

Also I looked at the source code, and it seems suspicious to me that `xray_index` is one of the only properties where no `hcl` annotation exists.

https://github.com/jfrog/terraform-provider-artifactory/blob/9530bd473a7267489f031cc27b08d8c47aee44ad/pkg/artifactory/resource/repository/remote/remote.go#L32

Also seems related to #494

**Comment by morgenthalep** [comment]
Template asks for fully functioning example, but since we use vault, I replaced those with placeholders.

I think it should work with pretty much any configuration though, since regardless of circumstance `xray_index` just isn't being set to anything other than false.

**Comment by alexhung** [comment]
@morgenthalep Re:

> I looked at the source code, and it seems suspicious to me that xray_index is one of the only properties where no hcl annotation exists.

That isn't the problem. We have code that use convention to infer the HCL attribute name. e.g. it infers the attribute name to be the snake_case of `XrayIndex`.

Can you confirm that the remote repo has the Xray Index set in Artifactory UI? Additionally, you can also use the API to get this value. i.e.

curl --request GET 'https://${var.artifactory_domain}/artifactory/api/repositories/docker-hub-remote'

After applying the change, the provider performs a READ from the Artifactory API to get the latest state. So if Artifactory does not toggle `xrayIndex` for this repo, then the provider state will reflect that (vs the value in your HCL).

**Comment by morgenthalep** [comment]
@alexhung Before the deploy I modified the terraform so it had no diffs, which included enabling xray_index for most repos, the only change was that I made username a sensitive property.
After the apply, all repositories with authentication (where there was any diff at all `username -> Sensistive`) suddenly had `xray_index` turned off.
This happened both on our test and prod environment, but I wrote it off as existing divergence on test.
I just had to re-enable `xray_index` on all our remote docker repos where we had it enabled, and after a refresh state, all diffs are gone in plan again.

**Comment by morgenthalep** [comment]
BTW, when I created this issue it didn't give the template selection for some reason.
It was intended to be a bug report.

**Comment by alexhung** [comment]
@morgenthalep Let me summaries this issue and please correct me:
1. You have an existing remote repo
2. You define the HCL for this remote repo in the .tf file with matching attribute values
3. You import the existing remote repo state into the TF state(?)
4. When you made some changes to the HCL and run `terraform apply` the end state of the repo doesn't have `xray_index` set to `true`? Or is the repo (on Artifactory) has the correct setting for Xray Indexing but incorrectly in TF HCL/state?

I'm trying to get my hands around your issue so I can identify the cause.

**Comment by morgenthalep** [comment]
> @morgenthalep Let me summaries this issue and please correct me:
> 
>     1. You have an existing remote repo
> 
>     2. You define the HCL for this remote repo in the .tf file with matching attribute values
> 
>     3. You import the existing remote repo state into the TF state(?)
> 
>     4. When you made some changes to the HCL and run `terraform apply` the end state of the repo doesn't have `xray_index` set to `true`? Or is the repo (on Artifactory) has the correct setting for Xray Indexing but incorrectly in TF HCL/state?
> 
> 
> I'm trying to get my hands around your issue so I can identify the cause.

Yep, thats about it:
1. Create repo manually, with xray_index set to true
2. Import repo state
3. Adjust repo until no diff, includes setting xray_index to true in `.tf`
4. Run terraform apply
5. Observe that xray_index is now turned off, and that there is now a diff for it
6. Turn xray_index on manually until xray_index goes away

I haven't manually tested these steps, but this is what we did.

**Comment by danielmkn** [comment]
@morgenthalep, I can't reproduce the problem.

1. Created 2 repositories in the UI - `docker-test-remote` and `docker-test-remote1` with Xray indexing enabled and timeout numbers changed on the Advanced tab to not have default values.

2. Added these two resources to the .tf file:
```
resource "artifactory_remote_docker_repository" "my-remote-docker" {
  key                            = "docker-test-remote"
  url                            = "https://registry-1.docker.io/"
}

resource "artifactory_remote_docker_repository" "my-remote-docker1" {
  key                            = "docker-test-remote1"
  url                            = "https://registry-1.docker.io/"
}
```
3. Run `terraform import artifactory_remote_docker_repository.my-remote-docker docker-test-remote` and `terraform import artifactory_remote_docker_repository.my-remote-docker1 docker-test-remote1`. 
**Note**: When importing, Terraform is making the call to Artifactory to get repo info, and I see `"xrayIndex": true,` (also, all my other non-default parameters are preserved) for both repos. This means all is good on the Artifactory side, and repos were created correctly.

4. Run `terraform plan`:
```
Terraform will perform the following actions:

  # artifactory_remote_docker_repository.my-remote-docker will be updated in-place
  ~ resource "artifactory_remote_docker_repository" "my-remote-docker" {
      ~ assumed_offline_period_secs           = 301 -> 300
        id                                    = "docker-test-remote"
      ~ property_sets                         = [
          - "artifactory",
        ]
      - proxy                                 = "testProxy" -> null
      ~ xray_index                            = true -> false
        # (29 unchanged attributes hidden)

        # (1 unchanged block hidden)
    }

  # artifactory_remote_docker_repository.my-remote-docker1 will be updated in-place
  ~ resource "artifactory_remote_docker_repository" "my-remote-docker1" {
      ~ assumed_offline_period_secs           = 301 -> 300
        id                                    = "docker-test-remote1"
      ~ property_sets                         = [
          - "artifactory",
        ]
      ~ xray_index                            = true -> false
        # (29 unchanged attributes hidden)

        # (1 unchanged block hidden)
    }

```
`xray_index` is false in the plan, because it's not set in my HCL yet, and the default is `false` in the schema.
The reason why only `assumed_offline_period_secs` attribute is shown here - all the rest of timeout attributes are `Computed`. 

5. Adding all the missing attributes to each repo:
```
resource "artifactory_remote_docker_repository" "my-remote-docker" {
  key                            = "docker-test-remote"
  url                            = "https://registry-1.docker.io/"
  assumed_offline_period_secs    = 301
  proxy                          = "testProxy"
  property_sets = [
    "artifactory"
  ]
  xray_index = true
}

resource "artifactory_remote_docker_repository" "my-remote-docker1" {
  key                            = "docker-test-remote1"
  url                            = "https://registry-1.docker.io/"
  assumed_offline_period_secs    = 301
  property_sets = [
    "artifactory"
  ]
  xray_index = true
}
```
Run `terraform plan`, no differences are shown. Run `terraform apply`:
```
artifactory_remote_docker_repository.my-remote-docker1: Refreshing state... [id=docker-test-remote1]
artifactory_remote_docker_repository.my-remote-docker: Refreshing state... [id=docker-test-remote]

No changes. Your infrastructure matches the configuration.

Terraform has compared your real infrastructure against your configuration and found no differences, so no changes are needed.

Apply complete! Resources: 0 added, 0 changed, 0 destroyed.
```

**Comment by danielmkn** [comment]
@morgenthalep please let me know if you can consistently reproduce the issue with all these manual steps.

**Comment by morgenthalep** [comment]
I think this might be because there is no difference at all, so the bug doesn't occur.
For me there was a small diff, because I made username sensitive.
For the repos where there was no diff at all, this issue didn't occur.
Sorry for the slightly wrong reproduction steps.

**Comment by securedirective** [comment]
For what it's worth, I'm also hitting this bug on a daily basis. Every repo that this provider touches gets reset to xrayIndex=false, so we've had to adjust our company process to run terraform repeatedly until it stops prompting for changes.

**Comment by danielmkn** [comment]
ok, @securedirective, can you please give me an example HCL? I'll retest it with more fields.

**Comment by danielmkn** [comment]
> 

the rest of the fields don't affect `xray_index` at all. The issue is only possible if the field is not set in HCL, then it will be overwritten with `false`. I'm open to a zoom meeting if you're willing to show me the issue when `xray_index` is set to `true` in the HCL and the actual value gets overwritten to `false`, because I can't reproduce it.

**Comment by danielmkn** [comment]
Ok, I see it finally. 
The issue was affecting several fields, not only `xray_index`. 
There is a Create and Update function for the repository resource. The **create** function uses a `PUT` Artifactory call and sets all attributes correctly. 
The **update** function uses a `POST` call and sets default values to the fields if these fields weren't changed by this upgrade call. That was a part of the legacy provider and now it's fixed.
@morgenthalep and @securedirective, thanks for raising this issue! The fix will be released today in 6.16.4.

**Comment by securedirective** [comment]
@danielmkn I was going to try to isolate the problem this evening. I'm glad to see you discovered the cause before I got that far. Looking forward to the fix.
